src/include/storage/dictionary_encoding_tile.h

```diff
@@ -0,0 +1,148 @@
+
+class Tuple;
+class TileGroup;
+class TileGroupHeader;
+class TupleIterator;
+class Tile;
+/**
+ * Represents a Tile.
+ *
+ * Tiles are only instantiated via TileFactory.
+ *
+ * NOTE: MVCC is implemented on the shared TileGroupHeader.
+ */
+class DictEncodedTile : public Tile {
+  friend class TileFactory;
+  friend class TupleIterator;
+  friend class TileGroupHeader;
+  friend class gc::GCManager;
+
+  DictEncodedTile() = delete;
+  DictEncodedTile(DictEncodedTile const &) = delete;
+
+ public:
+  // Tile creator
+  DictEncodedTile(BackendType backend_type, TileGroupHeader *tile_header,
+       const catalog::Schema &tuple_schema, TileGroup *tile_group,
+       int tuple_count);
+
+  virtual ~DictEncodedTile();
+
+  //===--------------------------------------------------------------------===//
+  // Operations
+  //===--------------------------------------------------------------------===//
+
+  /**
+   * Insert tuple at slot
+   * NOTE : No checks, must be at valid slot.
+   */
+	// this function is used before GetValueFast which use
+	// original schema, so return original schema
+  const catalog::Schema *GetSchema() const override { return &original_schema; }  ;
+
+  /**
+   * Returns value present at slot
+   */
+  type::Value GetValue(const oid_t tuple_offset, const oid_t column_id) override ;
+
+  /*
+   * Faster way to get value
+   * By amortizing schema lookups
+   */
+   // since we know that to use this function, we first get schema and extract
+   // info from schema then use this function, so we assume the column offset
+   // is the original column offset, therefore transform needed
+  type::Value GetValueFast(const oid_t tuple_offset, const size_t column_offset,
+                           const type::TypeId column_type,
+                           const bool is_inlined) override ;
+
+  /**
+   * Sets value at tuple slot.
+   */
+   // same as tile, no need to override
+//  void SetValue(const type::Value &value, const oid_t tuple_offset,
+//                const oid_t column_id);
+
+  /*
+   * Faster way to set value
+   * By amortizing schema lookups
+   */
+   // also same as tile
+//  void SetValueFast(const type::Value &value, const oid_t tuple_offset,
+//                    const size_t column_offset, const bool is_inlined,
+//                    const size_t column_length);
+
+
+
+  //===--------------------------------------------------------------------===//
+  // Dictionary Encoding
+  //===--------------------------------------------------------------------===//
+
+	inline bool GetDictEncoded() const { return is_dict_encoded; }
+
+	// given a tile, encode this tile in current tile
+	// when initializing this encoded tile, use original tile's schema
+	void DictEncode(Tile *tile);
+
+	// decode tile and return a new tile that contain the decoded data
+  Tile* DictDecode();
+
+ protected:
+  // is dictionary encoded
+  bool is_dict_encoded;
```

Comment:
Consider trailing underscore for class data members per Google C++ Style Guide. Emailing the Peloton dev group to see if this should be enforced going forward.